src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -1623,6 +1632,92 @@ private void addCommentVisibilityToTable(
+
+        // first get all possible giver recipient pairs
+        Map<String, Map<String, Set<String>>> completeGiverRecipientMap =
+                fqLogic.buildCompleteGiverRecipientMap(feedbackSession, relatedQuestionsMap.values(), courseRoster);
+
+        // remove the existing responses in those pairs
+        for (FeedbackResponseAttributes existingResponse : existingResponses) {
+            Map<String, Set<String>> currGiverRecipientMap =
+                    completeGiverRecipientMap.get(existingResponse.getFeedbackQuestionId());
+            if (!currGiverRecipientMap.containsKey(existingResponse.getGiver())) {
+                continue;
+            }
+            currGiverRecipientMap.get(existingResponse.getGiver()).remove(existingResponse.getRecipient());
+        }
+
+        List<FeedbackResponseAttributes> missingResponses = new ArrayList<>();
+        // build dummy responses
+        for (Map.Entry<String, Map<String, Set<String>>> currGiverRecipientMapEntry
+                : completeGiverRecipientMap.entrySet()) {
+            FeedbackQuestionAttributes correspondingQuestion =
+                    relatedQuestionsMap.get(currGiverRecipientMapEntry.getKey());
+            String questionId = correspondingQuestion.getId();
+
+            for (Map.Entry<String, Set<String>> giverRecipientEntry
+                    : currGiverRecipientMapEntry.getValue().entrySet()) {
+                // giver
+                String giverIdentifier = giverRecipientEntry.getKey();
+                CourseRoster.ParticipantInfo giverInfo = courseRoster.getInfoForIdentifier(giverIdentifier);
+
+                for (String recipientIdentifier : giverRecipientEntry.getValue()) {
+                    // recipient
+                    CourseRoster.ParticipantInfo recipientInfo = courseRoster.getInfoForIdentifier(recipientIdentifier);
+
+                    // skip responses not in current section
+                    if (section != null
+                            && !giverInfo.getSectionName().equals(section)
+                                    && !recipientInfo.getSectionName().equals(section)) {
+                        continue;
+                    }
+
+                    FeedbackResponseAttributes missingResponse =
+                            FeedbackResponseAttributes.builder(questionId, giverIdentifier, recipientIdentifier)
+                                    .withCourseId(feedbackSession.getCourseId())
+                                    .withFeedbackSessionName(feedbackSession.getFeedbackSessionName())
+                                    .withGiverSection(giverInfo.getSectionName())
+                                    .withRecipientSection(recipientInfo.getSectionName())
+                                    .withResponseDetails(new FeedbackTextResponseDetails(
+                                                    Const.INSTRUCTOR_FEEDBACK_RESULTS_MISSING_RESPONSE))
```

Comment:
Is it possible/not too troublesome to generate the response details object based on the question type?

Reply:
Actually it is hard. There is no "default value factory" in the backend (but there is one in the frontend)

Reply:
Then it's ok to let it be like this